gitreceived/gitreceived.go

```diff
@@ -302,19 +320,139 @@ func exitStatus(err error) (exitStatusMsg, error) {
 
 var cacheMtx sync.Mutex
 
-func ensureCacheRepo(path string) error {
+func ensureCacheRepo(tempDir, path string) error {
 	cacheMtx.Lock()
 	defer cacheMtx.Unlock()
 
-	cachePath := *repoPath + "/" + path
-	if _, err := os.Stat(cachePath); os.IsNotExist(err) {
-		os.MkdirAll(cachePath, 0755)
+	cachePath := tempDir + "/" + path
+
+	if !*useBlobstore {
+		if _, err := os.Stat(cachePath); os.IsNotExist(err) {
+			os.MkdirAll(cachePath, 0755)
+			cmd := exec.Command("git", "init", "--bare")
+			cmd.Dir = cachePath
+			err = cmd.Run()
+			if err != nil {
+				return err
+			}
+		}
+		return ioutil.WriteFile(cachePath+"/hooks/pre-receive", prereceiveHook, 0755)
+	}
+
+	os.MkdirAll(cachePath, 0755)
+	res, err := http.Get("http://blobstore.discoverd/cache/" + path + ".tar")
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode == 200 {
+		// cache hit
+		r := tar.NewReader(res.Body)
+		for {
+			header, err := r.Next()
+			if err != nil {
+				if err == io.EOF {
+					break
+				}
+				return err
+			}
+
+			filename := cachePath + "/" + header.Name
+
+			switch header.Typeflag {
+			case tar.TypeDir:
+				if err := os.MkdirAll(filename, os.FileMode(header.Mode)); err != nil {
+					return err
+				}
+				if err := os.Chmod(filename, os.FileMode(header.Mode)); err != nil {
+					return err
+				}
+			case tar.TypeReg, tar.TypeRegA:
+				// if the files are out of order, the dir might not exist yet
+				if err := os.MkdirAll(filepath.Dir(filename), os.FileMode(header.Mode)); err != nil {
+					return err
+				}
+				writer, err := os.OpenFile(filename, os.O_RDWR|os.O_CREATE|os.O_TRUNC, os.FileMode(header.Mode))
+				if err != nil {
+					return err
+				}
+				defer writer.Close()
+				io.Copy(writer, r)
+			default:
+			}
+		}
+	} else {
 		cmd := exec.Command("git", "init", "--bare")
 		cmd.Dir = cachePath
 		err = cmd.Run()
 		if err != nil {
 			return err
 		}
+		return ioutil.WriteFile(cachePath+"/hooks/pre-receive", prereceiveHook, 0755)
 	}
-	return ioutil.WriteFile(cachePath+"/hooks/pre-receive", prereceiveHook, 0755)
+	return nil
+}
+
+func uploadCache(tempDir, path string) error {
+	cacheMtx.Lock()
+	defer cacheMtx.Unlock()
+
+	cachePath := tempDir + "/" + path
+
+	r, w := io.Pipe()
+	tw := tar.NewWriter(w)
+
+	go func() {
+		defer w.Close()
+		if err := filepath.Walk(cachePath, func(path string, file os.FileInfo, err error) error {
```

Comment:
This and the untar should be split out into a utility package since they are generic.